File: backend/app/schemas/download_settings.py
"""
Pydantic schemas for user download quality settings.

These schemas define the request/response models for user-level download
preferences including video quality, audio-only mode, and audio bitrate.
"""

from typing import Optional

from pydantic import BaseModel
from pydantic import Field

from app.core.constants import AUDIO_QUALITY_OPTIONS
from app.core.constants import DEFAULT_AUDIO_ONLY
from app.core.constants import DEFAULT_AUDIO_QUALITY
from app.core.constants import DEFAULT_VIDEO_QUALITY
from app.core.constants import VIDEO_QUALITY_OPTIONS


class DownloadSettings(BaseModel):
    """Response schema for user download settings."""

    video_quality: str = Field(
        default=DEFAULT_VIDEO_QUALITY,
        description="Video quality preference for URL downloads",
    )
    audio_only: bool = Field(
        default=DEFAULT_AUDIO_ONLY,
        description="Download only audio (no video)",
    )
    audio_quality: str = Field(
        default=DEFAULT_AUDIO_QUALITY,
        description="Audio bitrate preference for audio-only downloads",
    )

    class Config:
        json_schema_extra = {
            "example": {
                "video_quality": "best",
                "audio_only": False,
                "audio_quality": "best",
            }
        }


class DownloadSettingsUpdate(BaseModel):
    """Request schema for updating user download settings. All fields optional."""

    video_quality: Optional[str] = Field(
        default=None,
        description="Video quality preference for URL downloads",
    )
    audio_only: Optional[bool] = Field(
        default=None,
        description="Download only audio (no video)",
    )
    audio_quality: Optional[str] = Field(
        default=None,
        description="Audio bitrate preference for audio-only downloads",
    )

    class Config:
        json_schema_extra = {
            "example": {
                "video_quality": "1080p",
                "audio_only": False,
                "audio_quality": "best",
            }
        }


class DownloadSystemDefaults(BaseModel):
    """Response schema for system-level download defaults and available options."""

    video_quality: str = Field(
        default=DEFAULT_VIDEO_QUALITY,
        description="Default video quality setting",
    )
    audio_only: bool = Field(
        default=DEFAULT_AUDIO_ONLY,
        description="Default audio-only setting",
    )
    audio_quality: str = Field(
        default=DEFAULT_AUDIO_QUALITY,
        description="Default audio quality setting",
    )
    available_video_qualities: dict[str, str] = Field(
        default=VIDEO_QUALITY_OPTIONS,
        description="Available video quality options (key -> display label)",
    )
    available_audio_qualities: dict[str, str] = Field(
        default=AUDIO_QUALITY_OPTIONS,
        description="Available audio quality options (key -> display label)",
    )

    class Config:
        json_schema_extra = {
            "example": {
                "video_quality": "best",
                "audio_only": False,
                "audio_quality": "best",
                "available_video_qualities": {
                    "best": "Best Available",
                    "1080p": "1080p (Full HD)",
                    "720p": "720p (HD)",
                },
                "available_audio_qualities": {
                    "best": "Best Available",
                    "320": "320 kbps",
                    "192": "192 kbps",
                },
            }
        }

File: frontend/src/lib/api/downloadSettings.ts
/**
 * API service for download quality user settings
 *
 * Manages user preferences for URL download quality including:
 * - Video quality selection (best, 4K, 1440p, 1080p, 720p, 480p, 360p)
 * - Audio-only mode
 * - Audio bitrate selection (best, 320kbps, 192kbps, 128kbps)
 */

import axiosInstance from '../axios';

/**
 * System-level download defaults and available options
 */
export interface DownloadSystemDefaults {
  video_quality: string;
  audio_only: boolean;
  audio_quality: string;
  available_video_qualities: Record<string, string>;
  available_audio_qualities: Record<string, string>;
}

/**
 * User download settings
 */
export interface DownloadSettings {
  video_quality: string;
  audio_only: boolean;
  audio_quality: string;
}

/**
 * Request payload for updating download settings
 */
export interface DownloadSettingsUpdate {
  video_quality?: string;
  audio_only?: boolean;
  audio_quality?: string;
}

/**
 * Response from reset endpoint
 */
export interface DownloadSettingsResetResponse {
  message: string;
  default_settings: DownloadSettings;
}

/**
 * Default download settings (client-side fallback)
 */
export const DEFAULT_DOWNLOAD_SETTINGS: DownloadSettings = {
  video_quality: 'best',
  audio_only: false,
  audio_quality: 'best',
};

/**
 * Get user's download settings
 */
export async function getDownloadSettings(): Promise<DownloadSettings> {
  const response = await axiosInstance.get('/user-settings/download');
  return response.data;
}

/**
 * Update user's download settings
 */
export async function updateDownloadSettings(
  settings: DownloadSettingsUpdate
): Promise<DownloadSettings> {
  const response = await axiosInstance.put('/user-settings/download', settings);
  return response.data;
}

/**
 * Reset download settings to system defaults
 */
export async function resetDownloadSettings(): Promise<DownloadSettingsResetResponse> {
  const response = await axiosInstance.delete('/user-settings/download');
  return response.data;
}

/**
 * Get system default values for download settings
 */
export async function getDownloadSystemDefaults(): Promise<DownloadSystemDefaults> {
  const response = await axiosInstance.get('/user-settings/download/system-defaults');
  return response.data;
}
